src/whitenoise/groups.rs

```diff
@@ -502,72 +498,247 @@ impl Whitenoise {
         Ok(())
     }
 
-    /// Uploads a group image to a Blossom server and returns the encrypted metadata.
+    /// Syncs group image cache if needed (smart, hash-based check)
+    ///
+    /// This method is called after processing welcomes and commits to proactively
+    /// cache group images. It only downloads if:
+    /// 1. The group has an image set
+    /// 2. The image_hash is not already cached
+    ///
+    /// This ensures images are ready before the UI needs them, while avoiding
+    /// redundant downloads.
+    ///
+    /// # Arguments
+    /// * `account` - The account viewing the group
+    /// * `group_id` - The MLS group ID
+    pub(crate) async fn sync_group_image_cache_if_needed(
+        &self,
+        account: &Account,
+        group_id: &GroupId,
+    ) -> Result<()> {
+        let group: mdk_core::prelude::group_types::Group;
+        {
+            // Get group data to check if it has an image
+            let mdk = Account::create_mdk(account.pubkey, &self.config.data_dir)?;
+            group = mdk
+                .get_group(group_id)
+                .map_err(WhitenoiseError::from)?
+                .ok_or(WhitenoiseError::GroupNotFound)?;
+        }
+
+        // Check if group has an image set
+        let (image_hash, image_key, image_nonce) =
+            match (group.image_hash, group.image_key, group.image_nonce) {
+                (Some(hash), Some(key), Some(nonce)) => (hash, key, nonce),
+                _ => return Ok(()), // No image set, nothing to do
+            };
+
+        // Download and cache the image
+        self.download_and_cache_group_image(
+            &account.pubkey,
+            group_id,
+            &image_hash,
+            &image_key,
+            &image_nonce,
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Downloads, decrypts, and caches a group image if not already cached
     ///
-    /// This method uses the MDK library's `prepare_group_image_for_upload` function to:
-    /// 1. Encrypt the image using ChaCha20-Poly1305 AEAD
-    /// 2. Generate a SHA256 hash of the encrypted data
-    /// 3. Derive a deterministic upload keypair for Blossom authentication
+    /// # Arguments
+    /// * `account_pubkey` - The account accessing the image
+    /// * `group_id` - The MLS group ID
+    /// * `image_hash` - SHA-256 hash of the encrypted image
+    /// * `image_key` - Encryption key
+    /// * `image_nonce` - Encryption nonce
+    ///
+    /// # Returns
+    /// Path to the cached image file
+    async fn download_and_cache_group_image(
```

Comment:
This method is way too large, it also seems out of place to be here in groups. It's doing a bunch of things I'd expect to be in `media_files` instead of `groups`